src/LevelZeroIOGroup.cpp

```diff
@@ -0,0 +1,862 @@
+        }
+        if (domain_idx < 0 ||
+            domain_idx >= m_platform_topo.num_domain(domain_type)) {
+            throw Exception("LevelZeroIOGroup::" + std::string(__func__) +
+                            ": domain_idx out of range.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        int result = -1;
+        bool is_found = false;
+        std::shared_ptr<control_s> control = m_control_available.at(
+                                             control_name).controls.at(domain_idx);
+
+        // Check if control was already pushed
+        for (size_t ii = 0; !is_found && ii < m_control_pushed.size(); ++ii) {
+            // same location means this control or its alias was already pushed
+            if (m_control_pushed[ii] == control) {
+                result = ii;
+                is_found = true;
+            }
+        }
+        if (!is_found) {
+            // If not pushed, add to pushed control
+            result = m_control_pushed.size();
+            m_control_pushed.push_back(control);
+        }
+
+        return result;
+    }
+
+    // Parse and update saved values for signals
+    void LevelZeroIOGroup::read_batch(void)
+    {
+        m_is_batch_read = true;
+        // TODO: if ENERGY and ENERGY_TIMESTAMP or ACTIVE_TIME and
+        //       ACTIVE_TIMESTAMP signals are called at the same domain
+        //       level, instead call the paired version to guarantee we get the
+        //       value with the correct timestamp
+    }
+
+    // Write all controls that have been pushed and adjusted
+    void LevelZeroIOGroup::write_batch(void)
+    {
+        for (auto &sv : m_control_available) {
+            for (unsigned int domain_idx = 0;
+                 domain_idx < sv.second.controls.size(); ++domain_idx) {
+                if (sv.second.controls.at(domain_idx)->m_is_adjusted) {
+                    write_control(sv.first, sv.second.domain_type, domain_idx,
+                                  sv.second.controls.at(domain_idx)->m_setting);
+                }
+            }
+        }
+    }
+
+    // Return the latest value read by read_batch()
+    double LevelZeroIOGroup::sample(int batch_idx)
+    {
+        if (batch_idx < 0 || batch_idx >= (int)m_signal_pushed.size()) {
+            throw Exception("LevelZeroIOGroup::" + std::string(__func__) +
+                            ": batch_idx " + std::to_string(batch_idx) +
+                            " out of range", GEOPM_ERROR_INVALID,
+                            __FILE__, __LINE__);
+
+        }
+        //Not strictly necessary, but keeping to enforce general
+        //flow of read_batch followed by sample.
+        if (!m_is_batch_read) {
+            throw Exception("LevelZeroIOGroup::" + std::string(__func__) +
+                            ": signal has not been read.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        return m_signal_pushed[batch_idx]->sample();
```

Comment:
This should return the cached value.  The value should only change after a call to read_batch().